### ISSUE TITLE
Taxonomy app

### ISSUE THREAD

**Issue by Ferdi** [mitodl/lore#8]
Provides the APIs for a) defining the taxonomy system, managing vocabularies and taxonomy terms and adding them for each repository and b) assigning taxonomy terms to learning objects.

See also https://goo.gl/gmpedb

**Comment by noisecapella** [comment]
Some initial thoughts on an API for taxonomy. Does this make sense given the data?

GET taxonomy/learning_object/1/vocabularies - List of vocabularies for a learning object
GET taxonomy/learning_object/1/terms - List of valid terms for a learning object

GET taxonomy/vocabulary/1 - vocabulary information
POST taxonomy/vocabulary - create new vocabulary
PUT taxonomy/vocabulary/1 - update existing vocabulary
DELETE taxonomy/vocabulary/1 - delete vocabulary

GET taxonomy/term/1 - Get information about term
POST taxonomy/term - create term
PUT taxonomy/term/1 - update existing term
DELETE taxonomy/term/1 - delete term

GET taxonomy/term/1/learning_object - get a list of learning objects for a particular term
POST taxonomy/term/1/learning_object/7 - create link between the two
DELETE taxonomy/term/1/learning_object/7 - delete link

GET taxonomy/vocabulary/1/types - List of valid learning object types for vocabulary (Or maybe included in taxonomy/vocabulary/1)?
POST taxonomy/vocabulary/1/type - append a new learning object type to that list
DELETE taxonomy/vocabulary/1/type/type_name_here (is type name guaranteed to fit here?)

**Comment by Ferdi** [comment]
It would be a good start on the REST api.  This should just be a  Django app that simply implements the models. What you have here would go in the REST app https://goo.gl/P3NsMd

**Comment by Ferdi** [comment]
By "API" , I meant, public methods like :

```
taxonomy_get_vocabularies(type)
```

will return all vocabularies that applies to  learning object of certain type (e.g. "problems")
You'd then use that in REST to do what you have above 

GET taxonomy/vocabulary/1/types - List of valid learning object types for vocabulary

**Comment by noisecapella** [comment]
Do these models make sense?

class Vocabulary(models.Model):
    vid = models.AutoField(primary_key=True)
    rid = models.ForeignKey("Repository")
    name = models.TextField()
    description = models.TextField()
    required = models.BooleanField()
    type = models.TextField()  # TODO: or enum? CharField?
    weight = models.IntegerField()  # TODO: or float?

class VocabularyUnitType(models.Model):
    vid = models.ForeignKey("Vocabulary")
    type = models.TextField()

class Term(models.Model):
    tid = models.AutoField(primary_key=True)
    vid = models.ForeignKey("Vocabulary")
    label = models.TextField()
    weight = models.IntegerField()

class LearningObjectTerm(models.Model):
    loid = models.ForeignKey("LearningObject")
    tid = models.ForeignKey("Term")

**Comment by Ferdi** [comment]
looks good.  weight - int.  type  - enum

**Comment by Ferdi** [comment]
Can we rename VocabularyUnitType to VocabularyObjectType or VocabularyLearningObjectType ?

The LearningObject table was previously called Unit. I renamed it but forgot to rename the vocabulary_unittype table

**Comment by carsongee** [comment]
nit: I would prefer to just let Django make primary key, and not add the v letter since it is redundant

**Comment by carsongee** [comment]
Same for all these models, just let the pk be implicit

**Comment by carsongee** [comment]
Also, the foreign keys should generally just be the model name, i.e. vocabulary for the Vocabulary foreign key, it makes using the ORM more intuitive, i.e.:

``` python
vocab_unit_type = VocabularyUnitType.objects.get(id=1)
print(vocab_unit_type.vocabulary.description)
```

**Comment by ShawnMilo** [comment]
Also, the foreign key names should not contain `id`. For example, `LearningObjectTerm` should have `learning_object` and `term`, not `loid` and `tid`.

Edit: GMTA, @carsongee

**Comment by carsongee** [comment]
LOL

**Comment by pdpinch** [comment]
Would this be an easier conversation to have on a PR, around the code?

**Comment by ShawnMilo** [comment]
@pdpinch Sure. Carson and I were trying to recommend some common Django conventions we're used to seeing.

**Comment by pdpinch** [comment]
Closed via #42
